### PR TITLE
[bugfix] Apex never instantiated.

### LINKED ISSUE #7271: Lightning + DDP + apex.amp fails to initialize
## 🐛 Bug
Runtime error when using DDP with apex.amp.

## To Reproduce
Replace with
```python
    trainer = pl.Trainer(
        max_epochs=1,
        gpus=1,
        num_nodes=1,
        accelerator='ddp',
        amp_backend='apex',
        amp_level='O3',
        precision=16,
        progress_bar_refresh_rate=20
    )
```
And export Boring Model as .py to your machine with apex installed.

The initialization of apex.amp lies at  https://github.com/PyTorchLightning/pytorch-lightning/blob/e272bea4dcf679d15ab836310c82193641e79778/pytorch_lightning/plugins/precision/apex_amp.py#L41, called at https://github.com/PyTorchLightning/pytorch-lightning/blob/e272bea4dcf679d15ab836310c82193641e79778/pytorch_lightning/trainer/trainer.py#L440 . However, the initialization of apex.amp requires the model to be at GPU side, where the transition https://github.com/PyTorchLightning/pytorch-lightning/blob/e272bea4dcf679d15ab836310c82193641e79778/pytorch_lightning/plugins/training_type/ddp.py#L272 
called at https://github.com/PyTorchLightning/pytorch-lightning/blob/e272bea4dcf679d15ab836310c82193641e79778/pytorch_lightning/trainer/trainer.py#L481 happens after init. Thus, the if-statement https://github.com/PyTorchLightning/pytorch-lightning/blob/e272bea4dcf679d15ab836310c82193641e79778/pytorch_lightning/plugins/precision/apex_amp.py#L50 ignores to performing apex configuration.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds a `dispatch` to accelerator as it is needed for properly instantiating Apex with ddp_spawn.

Fixes #7271

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by SeanNaren** [comment]
Just to ensure I understand from a high level, because the model is only moved to device after calling `connect` in the precision plugin, you've introduced `dispatch` that is called after `pre_dispatch` which allows the precision plugin to be called after the model has moved to the correct device? This fixes the Apex issue as the model has to be moved to CUDA first.

**Review by SeanNaren — APPROVED** [review]
Approved, but definitely would like to see @justusschock or @awaelchli comments!

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -107,6 +107,11 @@ def pre_dispatch(self, trainer: 'pl.Trainer') -> None:
             self.setup_optimizers(trainer)
         self.precision_plugin.pre_dispatch()
 
+    def dispatch(self, trainer: 'pl.Trainer') -> None:
+        """Hook to do something before the training/evaluation/prediction starts."""
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply pytorch_lightning/accelerators/accelerator.py:111]: @tchaton 
As the order of hooks being executed could be confusing.

[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/accelerators/accelerator.py:111]: i find the pre/dispatch/post confusing now :/

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 2]
```diff
@@ -34,24 +35,23 @@ def __init__(self, amp_level: str = "O2") -> None:
...
+
+    def dispatch(self, trainer: "pl.Trainer") -> None:
+        if not self._connected:
+            accelerator = trainer.accelerator
+            model, optimizers = self.configure_apex(accelerator.lightning_module, accelerator.optimizers)
+            self.reinit_scheduler_properties(optimizers, accelerator.lr_schedulers)
+            self.model = model
+            self.optimizers = optimizers
+            self._connected = True
+        return super().dispatch(trainer)
```
Comment (ananthsub) [inline pytorch_lightning/plugins/precision/apex_amp.py:54]: its not specific to apex. any optimizer that defines state which isn't lazily instantiated needs to handle the device move?

Comment (tchaton) [reply pytorch_lightning/plugins/precision/apex_amp.py:54]: Hey @ananthsub. Not sure to fully follow you :)

**Comment by tchaton** [comment]
Hey @ananthsub,

I will make another PR to rename them. `on_predispatch`, `on_dispatch`, etc..

Best,
T.C

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 3]
```diff
@@ -34,24 +35,23 @@ def __init__(self, amp_level: str = "O2") -> None:
...
-        if model.device.type != "cuda":
-            return model, optimizers, lr_schedulers
-        model, optimizers = self.configure_apex(amp, model, list(optimizers), self.amp_level)
-        self.reinit_scheduler_properties(optimizers, lr_schedulers)
-        return model, optimizers, lr_schedulers
+    def pre_dispatch(self) -> None:
+        return super().pre_dispatch()
+
+    def dispatch(self, trainer: "pl.Trainer") -> None:
+        if not self._connected:
```
Comment (awaelchli) [inline pytorch_lightning/plugins/precision/apex_amp.py:47]: can you quickly remind us why dispatch is called multiple times ( so you have the connected flag)?

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 4]
```diff
@@ -34,24 +35,23 @@ def __init__(self, amp_level: str = "O2") -> None:
...
+            accelerator = trainer.accelerator
+            model, optimizers = self.configure_apex(accelerator.lightning_module, accelerator.optimizers)
```
Comment (awaelchli) [inline pytorch_lightning/plugins/precision/apex_amp.py:49]:
```suggestion
            model, optimizers = self.configure_apex(trainer.lightning_module, accelerator.optimizers)
```
Let's access it through the trainer like we do for the other plugins?

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 5]
```diff
@@ -34,24 +35,23 @@ def __init__(self, amp_level: str = "O2") -> None:
...
+            self.reinit_scheduler_properties(optimizers, accelerator.lr_schedulers)
+            self.model = model
+            self.optimizers = optimizers
```
Comment (awaelchli) [inline pytorch_lightning/plugins/precision/apex_amp.py:52]: why do we need to set model and optimizers? The accelerator owns them. If you put the reference here and we are not careful, it gets out of sync / memory leaked.

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator.py:111]: Yes, we should think about the naming of these hooks. But more importantly, I think we can do a better job at formally _defining_ what these hook are supposed to do. Maybe another action item for 1.3 is to do a full pass over the plugins and improve all these docs. That would help everyone 1. implementing plugins 2. fix 3. review Plugin PRs.

File: tests/plugins/test_amp_plugins.py [hunk 6]
```diff
@@ -83,3 +97,37 @@ def test_amp_gradient_unscale(tmpdir, accum: int):
         accumulate_grad_batches=accum,
     )
     trainer.fit(model)
+
+
+@RunIf(min_gpus=2, amp_apex=True, special=True)
+@pytest.mark.parametrize("amp_level", ['O2'])
+def test_amp_apex_ddp_fit(amp_level, tmpdir):
+
```
Comment (awaelchli) [inline tests/plugins/test_amp_plugins.py:105]: We have another apex test in 
`tests/models/test_amp.py::test_amp_with_apex` It uses 1 gpu.
@tchaton your fix only applies to multi-gpu due to the dispatch, am I right?

**Review by awaelchli — APPROVED** [review]
cautious approval

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/plugins/precision/apex_amp.py:54]: @ananthsub I think this is handled by @awaelchli in #7277

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 4 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/precision/apex_amp.py:49]: @tchaton n00b question: would we be able to use for apex?

https://github.com/PyTorchLightning/pytorch-lightning/blob/338f5a331155d61a9d94168480ca45c264821019/pytorch_lightning/plugins/training_type/training_type_plugin.py#L193-L201

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 3 above)
Comment (justusschock) [reply pytorch_lightning/plugins/precision/apex_amp.py:47]: yes, I think this should only be called once.

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 5 above)
Comment (justusschock) [reply pytorch_lightning/plugins/precision/apex_amp.py:52]: can we make this a weaker if necessary at all? If possible we should really avoid setting them here, since this will get us back to the previous state where it was owned by everything

**Review by justusschock — APPROVED** [review]
Also carefully approving. I get that this is necessary, but I don't like simply assigning optimiser and model to every class :)

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/accelerators/accelerator.py:111]: n00b question: would this be easier if the precision plugin was owned by the training type plugin instead of the accelerator?

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/plugins/precision/apex_amp.py]: Done !

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/plugins/precision/apex_amp.py:47]: Yes, exactly.

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/plugins/precision/apex_amp.py]: I messed up. It should be attached to the training_type_plugin.

File: tests/plugins/test_amp_plugins.py (hunk 6 above)
Comment (tchaton) [reply tests/plugins/test_amp_plugins.py:105]: Yes.

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/accelerators/accelerator.py:111]: n00b question: would this be easier if the training type plugin maintained the precision plugin? then we could handle

[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator.py:111]: Yes, I think that could make it easier to interleave these operations between one plugin and the other. Here in this PR we see that the precision plugin needs to configure the model before it is wrapped, and needs to overwrite the reference in the training plugin. this really breaks the contract that these plugins currently have with each other.

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/precision/apex_amp.py:47]: ok so I understand now. it is called once within a fit, so that's good. but we don't teardown and .test() etc. later on we can't configure again. 

got it.

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 7]
```diff
@@ -34,24 +35,27 @@ def __init__(self, amp_level: str = "O2") -> None:
...
+        return super().pre_dispatch()
+
+    def dispatch(self, trainer: "pl.Trainer") -> None:
+        if not self._connected:
+            accelerator = trainer.accelerator
+            training_type_plugin = accelerator.training_type_plugin
+            lightning_module, optimizers = self.configure_apex(trainer.lightning_module, accelerator.optimizers)
+            self.reinit_scheduler_properties(optimizers, accelerator.lr_schedulers)
+            if not isinstance(training_type_plugin.model, LightningModule):
+                training_type_plugin.model.module.module = lightning_module
```
Comment (awaelchli) [inline pytorch_lightning/plugins/precision/apex_amp.py:53]: this unwrapping + setter is doomed to fail haha
But I honestly don't know a better way right now that's quick to fix. 
So I suggest to put a TODO here, we definitely want to look into it, maybe the reorganization of moving the precision inside the trianing plugin, we could return the apex model back to the trainnig plugin.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 7 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/precision/apex_amp.py:53]: @awaelchli @tchaton if this is a temporary fix, could we mark these as private methods so we don't need to go through a full deprecation cycle for `dispatch` later on?

File: CHANGELOG.md [hunk 8]
```diff
@@ -403,8 +403,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed `parameters_to_ignore` not properly set to DDPWrapper ([#7239](https://github.com/PyTorchLightning/pytorch-lightning/pull/7239))
 
 
-- Fixed parsing of `fast_dev_run=True` with the built-in `ArgumentParser` ([#7240](https://github.com/PyTorchLightning/pytorch-lightning/pull/7240))
+- Fixed `apex` not properly instantiated ([#7274](https://github.com/PyTorchLightning/pytorch-lightning/pull/7274))
+
```
Comment (ananthsub) [inline CHANGELOG.md:407]: is it for all `apex` use cases or specifically for ddp_spawn?

Comment (tchaton) [reply CHANGELOG.md]: Cleaned up !

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/plugins/precision/apex_amp.py]: Cleaned up !

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/accelerators/accelerator.py:111]: Yes, we should definitely refactor this and move optimizers, lr_schedulers to the training_type_plugin.
